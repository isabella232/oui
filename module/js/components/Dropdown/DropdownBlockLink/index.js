function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React, { useEffect, useRef } from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
export var DropdownBlockLink = function DropdownBlockLink(props) {
  var el = useRef(null);
  useEffect(function () {
    if (props.hasFauxFocus && el.current) {
      el.current.scrollIntoView({
        behavior: 'smooth',
        block: 'center'
      });
    }
  });

  function onClick() {
    props.onClick(props.value);
  }

  var styleProps = {};

  if (props.minWidth) {
    styleProps.minWidth = props.minWidth;
  }

  return React.createElement("div", _extends({
    className: classNames({
      'link': props.isLink,
      'isSelected': !props.isLink,
      'oui-dropdown__block-link': props.isLink,
      'oui-dropdown__block-link--has-focus': props.hasFauxFocus
    }),
    ref: el,
    style: styleProps
  }, props.testSection ? {
    'data-test-section': props.testSection
  } : {}, props.trackId ? {
    'data-track-id': props.trackId
  } : {}, {
    onClick: onClick
  }), props.children);
};
DropdownBlockLink.defaultProps = {
  isLink: true
};
DropdownBlockLink.propTypes = {
  /** Content to be shown in the menu option */
  children: PropTypes.node.isRequired,

  /** Whether or not this element should be highlighted. */
  hasFauxFocus: PropTypes.bool,

  /** should this be a link or nonclickable text */
  isLink: PropTypes.bool.isRequired,

  /** Minimum width of the list item, useful if
    * you need to have a block of description text */
  minWidth: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),

  /** Click handler for the menu option */
  onClick: PropTypes.func.isRequired,

  /** Used for data-test-section attribute on the link */
  testSection: PropTypes.string,

  /** Used for data-track-id attribute on the link */
  trackId: PropTypes.string,

  /** An optional value to invoke the onClick callback with */
  value: PropTypes.string
};
export default DropdownBlockLink;