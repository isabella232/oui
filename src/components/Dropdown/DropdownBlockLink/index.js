import React, { useEffect, useRef } from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import Checkbox from '../../Checkbox';

export const DropdownBlockLink = (props) => {
  const el = useRef(null);

  useEffect(() => {
    if (props.hasFauxFocus && el.current) {
      el.current.scrollIntoView({
        behavior: 'smooth',
        block: 'center',
      });
    }
  });

  function onClick() {
    props.onClick(props.value);
  }

  const styleProps = {};
  if (props.minWidth) {
    styleProps.minWidth = props.minWidth;
  }

  return (
    <div
      className={ classNames({
        'link': props.isLink,
        'isSelected': !props.isLink,
        'oui-dropdown__block-link': props.isLink,
        'oui-dropdown__block-link--has-focus': props.hasFauxFocus,
      }) }
      ref={ el }
      style={ styleProps }
      { ...(props.testSection ? { 'data-test-section': props.testSection } : {}) }
      { ...(props.trackId ? { 'data-track-id': props.trackId } : {}) }
      onClick={ onClick }>
      {
        props.isMultiSelect ?
          <Checkbox
            defaultChecked={ props.isItemSelected }
            label={ props.children }
            isDisabled={ false }
          /> :
          props.children
      }
    </div>
  );
};

DropdownBlockLink.defaultProps = {
  isLink: true,
};

DropdownBlockLink.propTypes = {
  /** Content to be shown in the menu option */
  children: PropTypes.node.isRequired,
  /** Whether or not this element should be highlighted. */
  hasFauxFocus: PropTypes.bool,
  /** Whether or not this item's checkbox is checked' */
  isItemSelected: PropTypes.bool,
  /**
   * Should this be a link or non-clickable text?
   * To indicate item is selected (in single select),
   * isLink should be false.
   */
  isLink: PropTypes.bool.isRequired,
  /** Whether or not this item should include a checkbox */
  isMultiSelect: PropTypes.bool,
  /** Minimum width of the list item, useful if
    * you need to have a block of description text */
  minWidth: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /** Click handler for the menu option */
  onClick: PropTypes.func.isRequired,
  /** Used for data-test-section attribute on the link */
  testSection: PropTypes.string,
  /** Used for data-track-id attribute on the link */
  trackId: PropTypes.string,
  /** An optional value to invoke the onClick callback with */
  value: PropTypes.string,
};

export default DropdownBlockLink;
