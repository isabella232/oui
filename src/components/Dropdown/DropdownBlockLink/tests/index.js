import React from 'react';
import DropdownBlockLink from '../index';
import { shallow, mount } from 'enzyme';
import { mountToJson } from 'enzyme-to-json';


describe('components/Dropdown/DropdownBlockLink', () => {
  it('should render as a `div`', () => {
    const component = shallow(<DropdownBlockLink>foo</DropdownBlockLink>);
    expect(component.type()).toBe('div');
  });

  it('should render children', () => {
    const component = shallow(<DropdownBlockLink>foo</DropdownBlockLink>);
    expect(component.text()).toBe('foo');
  });

  it('should render with test section', () => {
    const component = shallow(<DropdownBlockLink testSection="goose">foo</DropdownBlockLink>);
    expect(component.is('[data-test-section="goose"]')).toBe(true);
  });

  it('should render with a track id', () => {
    const component = shallow(<DropdownBlockLink trackId="goose">foo</DropdownBlockLink>);
    expect(component.is('[data-track-id="goose"]')).toBe(true);
  });

  it('should render with a min width', () => {
    const component = shallow(<DropdownBlockLink minWidth="300px">foo</DropdownBlockLink>);
    expect(component.prop('style')).toEqual({ minWidth: '300px' });
  });

  it('should render with link styling by default', () => {
    const component = shallow(<DropdownBlockLink>foo</DropdownBlockLink>);
    expect(component.prop('className')).toEqual('link oui-dropdown__block-link');
  });

  it('should render with link styling if prop passed as true', () => {
    const component = shallow(<DropdownBlockLink isLink={ true }>foo</DropdownBlockLink>);
    expect(component.prop('className')).toEqual('link oui-dropdown__block-link');
  });

  it('should render with link styling if prop passed as false', () => {
    const component = shallow(<DropdownBlockLink isLink={ false }>foo</DropdownBlockLink>);
    expect(component.prop('className')).toEqual('isSelected');
  });

  it('should call function that is passed in as `onClick` after click', () => {
    const obj = { func: () => {} };
    spyOn(obj, 'func').and.stub();

    const component = shallow(<DropdownBlockLink onClick={ obj.func }>Hello!</DropdownBlockLink>);

    component.simulate('click');

    expect(obj.func).toHaveBeenCalled();
  });

  it('should add a checkbox when isMultiSelect is true', () => {
    const component = shallow(<DropdownBlockLink isMultiSelect={ true }>foo</DropdownBlockLink>);
    expect(component.find('Checkbox').length).toBe(1);
  });

  it('should add a checked checkbox when isItemSelected is true', () => {
    const component = mount(<DropdownBlockLink isItemSelected={ true } isMultiSelect={ true }>foo</DropdownBlockLink>);
    expect(mountToJson(component)).toMatchSnapshot();
  });
});
