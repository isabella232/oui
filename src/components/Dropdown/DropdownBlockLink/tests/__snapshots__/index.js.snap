// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Dropdown/DropdownBlockLink should add a checked checkbox when isItemSelected is true 1`] = `
<DropdownBlockLink
  isItemSelected={true}
  isLink={true}
  isMultiSelect={true}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="link oui-dropdown__block-link"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={Object {}}
  >
    <Checkbox
      defaultChecked={true}
      isDisabled={false}
      label="foo"
      labelWeight="normal"
      onChange={[Function]}
    >
      <Label>
        <label
          className={null}
          data-oui-component={true}
        >
          <div
            className="flex"
          >
            <input
              className="flex--none"
              data-oui-component={true}
              defaultChecked={true}
              disabled={false}
              onChange={[Function]}
              style={
                Object {
                  "marginTop": "0.35em",
                }
              }
              type="checkbox"
            />
            <div
              className="flush--bottom push--left weight--normal cursor--pointer"
            >
              foo
            </div>
          </div>
        </label>
      </Label>
    </Checkbox>
  </div>
</DropdownBlockLink>
`;
